File: 57-insert-interval/57-insert-interval.js
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
const insert = function(intervals, newInterval) {
    const result = [];
    
    for (let i = 0, n = intervals.length; i < n; i++) {
        if (newInterval[1] < intervals[i][0]) {
            result.push(newInterval);
            return result.concat(intervals.slice(i));
        }
        
        if (newInterval[0] > intervals[i][1]) {
            result.push(intervals[i]);
            continue;
        }
        
        const newStart = Math.min(newInterval[0], intervals[i][0]);
        const newEnd = Math.max(newInterval[1], intervals[i][1]);
        newInterval = [newStart, newEnd];
    }
    result.push(newInterval); 
    return result;
};

File: 57-insert-interval/57-insert-interval.rb
# @param {Integer[][]} intervals
# @param {Integer[]} new_interval
# @return {Integer[][]}
def insert(intervals, new_interval)
    return [new_interval] if intervals.empty? 
    result = []
    interval_to_add = new_interval
    
    intervals.each do |interval|
        if interval[1] < new_interval[0] || interval_to_add.nil?
            result.push(interval)
        elsif interval_to_add[0] <= interval[1] && interval_to_add[1] >= interval[0] 
            interval_to_add = [[interval_to_add[0], interval[0]].min, [interval_to_add[1], interval[1]].max]
        else
            result.push(interval_to_add, interval)
            interval_to_add = nil
        end
    end
    result.push(interval_to_add) if interval_to_add
    result
end